File: golang/12_run_evaluation.go
// Recipe 12: run an evaluation as a 4-level Call trace.
//
// The integration recipe. Looks up everything earlier recipes created,
// builds the structured Call tree the W&B UI recognises as an evaluation
// run, and verifies via /eval_results/query. Lands ADR-0005 (the
// imperative-SDK-path decision).
//
// The trace shape mirrors the SDK's evaluation.evaluate(model):
//
//   Evaluation.evaluate                  (root, op_name = canonical ref)
//   |-- Evaluation.predict_and_score     (per-row trial)
//   |   |-- <Model>.predict              (the model invocation)
//   |   `-- <scorer>                     (scoring)
//   |-- ... (one predict_and_score per row)
//   `-- Evaluation.summarize             (sibling of predict_and_score)
//
// This recipe *creates only Calls* — recipe 11 owns the eval's definition
// (Eval Object + canonical Ops); recipe 08 owns the Model + predict Op.
//
// Wire-level points worth knowing:
//
//   - Per-Call op_name MUST be a weave:// URI to an existing Op.
//   - The root Call's display_name is the Evaluations-page label; without
//     it every run shows the bare op_name. Set to eval-golang-<unix>.
//   - Root /call/end summary needs weave.status="success" and
//     status_counts.success = total call count (1 + N*3 + 1).
//   - The per-row `scores` key and the summarize/root output keys MUST be
//     the scorer Op's short name (object_id) — that's what links per-row
//     scorer_key back to the Eval Object's val.scorers and powers the
//     leaderboard. The model invocation is mocked (always returns the
//     expected answer), so pass_rate is 1.0.
//
// Run:
//   go run golang/12_run_evaluation.go

package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
	"strings"
	"time"
)

const modelLatency = 0.001

var (
	entity    = os.Getenv("WANDB_ENTITY")
	project   = os.Getenv("WANDB_PROJECT")
	projectID = entity + "/" + project
	baseURL   = getenv("WEAVE_SERVICE_URL", "https://trace.wandb.ai")
	apiKey    = os.Getenv("WANDB_API_KEY")
	client    = &http.Client{}

	attributes = map[string]any{
		"cookbook.language":    "golang",
		"cookbook.recipe":      "12_run_evaluation",
		"cookbook.environment": getenv("COOKBOOK_ENVIRONMENT", "dev"),
	}
	datasetRefRe = regexp.MustCompile(`weave:///[^/]+/[^/]+/object/([^:]+):(.+)`)
	tableRefRe   = regexp.MustCompile(`/table/([A-Za-z0-9_-]+)$`)
)

func getenv(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

func fatal(format string, args ...any) {
	fmt.Fprintf(os.Stderr, format+"\n", args...)
	os.Exit(1)
}

func now() string { return time.Now().UTC().Format(time.RFC3339Nano) }

func do(method, path string, body map[string]any) map[string]any {
	var reader io.Reader
	if body != nil {
		payload, err := json.Marshal(body)
		if err != nil {
			fatal("encode %s: %v", path, err)
		}
		reader = bytes.NewReader(payload)
	}
	req, err := http.NewRequest(method, baseURL+path, reader)
	if err != nil {
		fatal("request %s: %v", path, err)
	}
	req.SetBasicAuth("api", apiKey)
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	res, err := client.Do(req)
	if err != nil {
		fatal("%s %s: %v", method, path, err)
	}
	defer res.Body.Close()
	raw, _ := io.ReadAll(res.Body)
	if res.StatusCode/100 != 2 {
		fatal("HTTP %d for %s %s: %s", res.StatusCode, method, path, raw)
	}
	if len(raw) == 0 {
		return map[string]any{}
	}
	var out map[string]any
	if err := json.Unmarshal(raw, &out); err != nil {
		fatal("decode %s: %v", path, err)
	}
	return out
}

func post(path string, body map[string]any) map[string]any { return do(http.MethodPost, path, body) }
func get(path string) map[string]any                       { return do(http.MethodGet, path, nil) }

func latestObject(objectID string) map[string]any {
	r := post("/objs/query", map[string]any{
		"project_id":    projectID,
		"filter":        map[string]any{"object_ids": []string{objectID}, "latest_only": true},
		"metadata_only": false,
	})
	objs, _ := r["objs"].([]any)
	if len(objs) == 0 {
		return nil
	}
	o, _ := objs[0].(map[string]any)
	return o
}

// startCall opens a Call; parentID/traceID/displayName are omitted when "".
func startCall(opName string, inputs map[string]any, parentID, traceID, displayName string) (string, string) {
	start := map[string]any{
		"project_id": projectID,
		"op_name":    opName,
		"started_at": now(),
		"attributes": attributes,
		"inputs":     inputs,
	}
	if parentID != "" {
		start["parent_id"] = parentID
	}
	if traceID != "" {
		start["trace_id"] = traceID
	}
	if displayName != "" {
		start["display_name"] = displayName
	}
	r := post("/call/start", map[string]any{"start": start})
	return r["id"].(string), r["trace_id"].(string)
}

// endCall closes a Call with the default success summary.
func endCall(callID string, output any) {
	post("/call/end", map[string]any{"end": map[string]any{
		"project_id": projectID,
		"id":         callID,
		"ended_at":   now(),
		"summary": map[string]any{
			"status_counts": map[string]any{"success": 1, "error": 0},
			"weave":         map[string]any{"status": "success"},
		},
		"output": output,
	}})
}

func main() {
	var missing []string
	for _, k := range []string{"WANDB_API_KEY", "WANDB_ENTITY", "WANDB_PROJECT"} {
		if os.Getenv(k) == "" {
			missing = append(missing, k)
		}
	}
	if len(missing) > 0 {
		fatal("Missing required env vars: %s. See ../README.md#setup.", strings.Join(missing, ", "))
	}

	// 1) Look up the Evaluation Object + extract refs from its val.
	evalObj := latestObject("recipe-11-eval-golang")
	if evalObj == nil {
		fatal("FAIL: Evaluation Object `recipe-11-eval-golang` not found. Run golang/11_create_evaluation.go first.")
	}
	evalObjRef := fmt.Sprintf("weave:///%s/object/%s:%s", projectID, evalObj["object_id"], evalObj["digest"])
	ev, _ := evalObj["val"].(map[string]any)
	evaluateOpRef := ev["evaluate"].(string)
	predictAndScoreOpRef := ev["predict_and_score"].(string)
	summarizeOpRef := ev["summarize"].(string)
	scorers, _ := ev["scorers"].([]any)
	scorerOpRef := scorers[0].(string)
	datasetRef := ev["dataset"].(string)
	// The scorer Op's short_name (object_id) keys the per-row scores, the
	// wandb.runnable.* feedback_type, and the summarize/root output.
	afterOp := scorerOpRef[strings.LastIndex(scorerOpRef, "/op/")+len("/op/"):]
	scorerShortName := strings.SplitN(afterOp, ":", 2)[0]
	fmt.Printf("Eval obj:  %v digest=%v…\n", evalObj["object_id"], evalObj["digest"].(string)[:12])

	// 2) Look up the Model + its predict Op (recipe 08).
	modelObj := latestObject("recipe-08-model-golang")
	if modelObj == nil {
		fatal("FAIL: Model `recipe-08-model-golang` not found. Run golang/08_use_model.go first.")
	}
	modelRef := fmt.Sprintf("weave:///%s/object/%s:%s", projectID, modelObj["object_id"], modelObj["digest"])
	modelPredictOp := latestObject("recipe-08-model-golang.predict")
	if modelPredictOp == nil {
		fatal("FAIL: Model predict Op `recipe-08-model-golang.predict` not found. Run golang/08_use_model.go first.")
	}
	modelPredictOpRef := fmt.Sprintf("weave:///%s/op/%s:%s", projectID, modelPredictOp["object_id"], modelPredictOp["digest"])
	fmt.Printf("Model:     %v digest=%v…\n", modelObj["object_id"], modelObj["digest"].(string)[:12])

	// 3) Walk the Dataset rows.
	m := datasetRefRe.FindStringSubmatch(datasetRef)
	if m == nil {
		fatal("FAIL: could not parse dataset_ref: %q", datasetRef)
	}
	dsID, dsDigest := m[1], m[2]
	dsMeta := get(fmt.Sprintf("/v2/%s/%s/datasets/%s/versions/%s", entity, project, dsID, dsDigest))
	rowsRef := dsMeta["rows"].(string)
	tableDigest := rowsRef
	if tm := tableRefRe.FindStringSubmatch(rowsRef); tm != nil {
		tableDigest = tm[1]
	}
	rowsRes := post("/table/query", map[string]any{"project_id": projectID, "digest": tableDigest})
	rawRows, _ := rowsRes["rows"].([]any)
	var rows []map[string]any
	for _, rr := range rawRows {
		if r, ok := rr.(map[string]any); ok {
			rows = append(rows, r["val"].(map[string]any))
		}
	}
	fmt.Printf("Dataset:   %s (%d rows)\n", dsID, len(rows))

	// 4) Build the 4-level Call trace. The display_name on the root is the
	// Evaluations-page label.
	displayName := fmt.Sprintf("eval-golang-%d", time.Now().Unix())
	rootID, traceID := startCall(evaluateOpRef,
		map[string]any{"self": evalObjRef, "model": modelRef}, "", "", displayName)
	fmt.Printf("Root call: %s (display_name=%q)\n", rootID, displayName)

	nPass := 0
	totalCalls := 1 // root
	for _, row := range rows {
		psID, _ := startCall(predictAndScoreOpRef,
			map[string]any{"self": evalObjRef, "model": modelRef, "example": row}, rootID, traceID, "")

		// Predict child: invoke the (mocked) model — always returns expected.
		predID, _ := startCall(modelPredictOpRef,
			map[string]any{"self": modelRef, "question": row["question"]}, psID, traceID, "")
		prediction := row["answer"]
		endCall(predID, map[string]any{"answer": prediction})

		// Scorer child: compare prediction vs expected.
		scID, _ := startCall(scorerOpRef,
			map[string]any{"output": prediction, "expected": row["answer"]}, psID, traceID, "")
		score := prediction == row["answer"]
		endCall(scID, score)

		// Link the score to the predict Call via a wandb.runnable.* Feedback
		// row (recipe 09's pattern) so the leaderboard attributes the scorer Op.
		predCallRef := fmt.Sprintf("weave:///%s/call/%s", projectID, predID)
		scoreCallRef := fmt.Sprintf("weave:///%s/call/%s", projectID, scID)
		post("/feedback/create", map[string]any{
			"project_id":    projectID,
			"weave_ref":     predCallRef,
			"feedback_type": "wandb.runnable." + scorerShortName,
			"payload":       map[string]any{"output": score},
			"runnable_ref":  scorerOpRef,
			"call_ref":      scoreCallRef,
		})

		// End predict_and_score. The `scores` key MUST be the scorer Op's
		// short name so /eval_results/query links it to val.scorers.
		endCall(psID, map[string]any{
			"output":        prediction,
			"scores":        map[string]any{scorerShortName: score},
			"model_latency": modelLatency,
		})

		if score {
			nPass++
		}
		totalCalls += 3 // predict_and_score + predict + scorer
	}

	// Summarize: sibling of predict_and_score under the root.
	sumID, _ := startCall(summarizeOpRef, map[string]any{"self": evalObjRef}, rootID, traceID, "")
	passRate := 0.0
	if len(rows) > 0 {
		passRate = float64(nPass) / float64(len(rows))
	}
	// summarize.output AND root.output are keyed by the scorer short name
	// (matching val.scorers + the per-row scorer_key) plus model_latency.mean
	// — this is what the leaderboard view buckets across runs.
	aggregatedOutput := map[string]any{
		scorerShortName: map[string]any{"true_count": nPass, "true_fraction": passRate},
		"model_latency": map[string]any{"mean": modelLatency},
	}
	endCall(sumID, aggregatedOutput)
	totalCalls++ // summarize

	// 5) End the root with the proper summary shape.
	post("/call/end", map[string]any{"end": map[string]any{
		"project_id": projectID,
		"id":         rootID,
		"ended_at":   now(),
		"summary": map[string]any{
			"status_counts": map[string]any{"success": totalCalls, "error": 0},
			"weave":         map[string]any{"status": "success", "display_name": displayName},
		},
		"output": aggregatedOutput,
	}})
	fmt.Printf("Trace done: %d calls, pass_rate=%.2f\n", totalCalls, passRate)

	// --- verification ---
	// /eval_results/query with the root call_id aggregates per-row trial data
	// + scorer stats. The summary's evaluation_ref should match our Eval Object.
	time.Sleep(2 * time.Second)
	var results map[string]any
	for i := 0; i < 8; i++ {
		r := post(fmt.Sprintf("/v2/%s/%s/eval_results/query", entity, project), map[string]any{
			"evaluation_call_ids": []string{rootID},
			"include_rows":        true,
			"include_summary":     true,
		})
		if tr, ok := r["total_rows"].(float64); ok && int(tr) == len(rows) {
			results = r
			break
		}
		time.Sleep(time.Second)
	}
	if results == nil {
		fatal("FAIL: eval_results/query did not return %d rows after 8 attempts", len(rows))
	}

	summary, _ := results["summary"].(map[string]any)
	evals, _ := summary["evaluations"].([]any)
	if len(evals) != 1 {
		fatal("expected 1 evaluation in summary, got %d", len(evals))
	}
	evSummary, _ := evals[0].(map[string]any)
	if evSummary["evaluation_ref"] != evalObjRef {
		fatal("evaluation_ref: %v", evSummary["evaluation_ref"])
	}
	var scorerKeys []string
	stats, _ := evSummary["scorer_stats"].([]any)
	for _, s := range stats {
		if sm, ok := s.(map[string]any); ok {
			if k, ok := sm["scorer_key"].(string); ok {
				scorerKeys = append(scorerKeys, k)
			}
		}
	}
	found := false
	for _, k := range scorerKeys {
		if k == scorerShortName {
			found = true
		}
	}
	if !found {
		fatal("%q missing from scorer_stats: %v", scorerShortName, scorerKeys)
	}
	fmt.Printf("Verified:  /eval_results/query returned %v rows, evaluation_ref matches, scorer_stats=%v\n", results["total_rows"], scorerKeys)
}

File: java/12_RunEvaluation.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 17+
//DEPS com.fasterxml.jackson.core:jackson-databind:2.18.2

// Recipe 12: run an evaluation as a 4-level Call trace.
//
// The integration recipe. Looks up everything earlier recipes created,
// builds the structured Call tree the W&B UI recognises as an evaluation
// run, and verifies via /eval_results/query. Lands ADR-0005 (the
// imperative-SDK-path decision).
//
// The trace shape mirrors the SDK's evaluation.evaluate(model):
//
//   Evaluation.evaluate                  (root, op_name = canonical ref)
//   |-- Evaluation.predict_and_score     (per-row trial)
//   |   |-- <Model>.predict              (the model invocation)
//   |   `-- <scorer>                     (scoring)
//   |-- ... (one predict_and_score per row)
//   `-- Evaluation.summarize             (sibling of predict_and_score)
//
// This recipe *creates only Calls* — recipe 11 owns the eval's definition
// (Eval Object + canonical Ops); recipe 08 owns the Model + predict Op.
//
// Wire-level points worth knowing:
//
//   - Per-Call op_name MUST be a weave:// URI to an existing Op.
//   - The root Call's display_name is the Evaluations-page label; without
//     it every run shows the bare op_name. Set to eval-java-<unix>.
//   - Root /call/end summary needs weave.status="success" and
//     status_counts.success = total call count (1 + N*3 + 1).
//   - The per-row `scores` key and the summarize/root output keys MUST be
//     the scorer Op's short name (object_id) — that's what links per-row
//     scorer_key back to the Eval Object's val.scorers and powers the
//     leaderboard. The model invocation is mocked (always returns the
//     expected answer), so pass_rate is 1.0.
//
// Run:
//   jbang java/12_RunEvaluation.java

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class RunEvaluation {
    static final ObjectMapper MAPPER = new ObjectMapper();
    static final HttpClient HTTP = HttpClient.newHttpClient();
    static final String BASE_URL = getenv("WEAVE_SERVICE_URL", "https://trace.wandb.ai");
    static final double MODEL_LATENCY = 0.001;
    static final Pattern DATASET_REF = Pattern.compile("weave:///[^/]+/[^/]+/object/([^:]+):(.+)");
    static final Pattern TABLE_REF = Pattern.compile("/table/([A-Za-z0-9_-]+)$");
    static String entity;
    static String project;
    static String projectId;
    static String authHeader;
    static Map<String, Object> attributes;

    static String getenv(String key, String def) {
        String v = System.getenv(key);
        return (v == null || v.isEmpty()) ? def : v;
    }

    static void fail(String msg) {
        System.err.println("FAIL: " + msg);
        System.exit(1);
    }

    static String now() { return Instant.now().toString(); }

    static JsonNode doReq(String method, String path, Object body) throws Exception {
        HttpRequest.Builder b = HttpRequest.newBuilder(URI.create(BASE_URL + path))
                .header("Authorization", authHeader);
        if (body != null) {
            b.header("Content-Type", "application/json")
                    .method(method, HttpRequest.BodyPublishers.ofString(MAPPER.writeValueAsString(body), StandardCharsets.UTF_8));
        } else {
            b.method(method, HttpRequest.BodyPublishers.noBody());
        }
        HttpResponse<String> res = HTTP.send(b.build(), HttpResponse.BodyHandlers.ofString());
        if (res.statusCode() / 100 != 2)
            fail("HTTP " + res.statusCode() + " for " + method + " " + path + ": " + res.body());
        String rb = res.body();
        return (rb == null || rb.isEmpty()) ? MAPPER.createObjectNode() : MAPPER.readTree(rb);
    }

    static JsonNode post(String path, Object body) throws Exception { return doReq("POST", path, body); }
    static JsonNode get(String path) throws Exception { return doReq("GET", path, null); }

    static JsonNode latestObject(String objectId) throws Exception {
        JsonNode r = post("/objs/query", Map.of(
                "project_id", projectId,
                "filter", Map.of("object_ids", List.of(objectId), "latest_only", true),
                "metadata_only", false));
        JsonNode objs = r.path("objs");
        return objs.isArray() && objs.size() > 0 ? objs.get(0) : null;
    }

    // Opens a Call; returns {id, trace_id}. parentId/traceId/displayName are
    // omitted when null.
    static String[] startCall(String opName, Map<String, Object> inputs, String parentId, String traceId, String displayName) throws Exception {
        var start = new LinkedHashMap<String, Object>();
        start.put("project_id", projectId);
        start.put("op_name", opName);
        start.put("started_at", now());
        start.put("attributes", attributes);
        start.put("inputs", inputs);
        if (parentId != null) start.put("parent_id", parentId);
        if (traceId != null) start.put("trace_id", traceId);
        if (displayName != null) start.put("display_name", displayName);
        JsonNode r = post("/call/start", Map.of("start", start));
        return new String[]{r.get("id").asText(), r.get("trace_id").asText()};
    }

    // Closes a Call with the default success summary.
    static void endCall(String callId, Object output) throws Exception {
        post("/call/end", Map.of("end", Map.of(
                "project_id", projectId,
                "id", callId,
                "ended_at", now(),
                "summary", Map.of(
                        "status_counts", Map.of("success", 1, "error", 0),
                        "weave", Map.of("status", "success")),
                "output", output)));
    }

    public static void main(String[] args) throws Exception {
        var missing = new ArrayList<String>();
        for (String k : List.of("WANDB_API_KEY", "WANDB_ENTITY", "WANDB_PROJECT"))
            if (getenv(k, "").isEmpty()) missing.add(k);
        if (!missing.isEmpty())
            fail("Missing required env vars: " + String.join(", ", missing) + ". See ../README.md#setup.");

        entity = System.getenv("WANDB_ENTITY");
        project = System.getenv("WANDB_PROJECT");
        projectId = entity + "/" + project;
        authHeader = "Basic " + Base64.getEncoder().encodeToString(
                ("api:" + System.getenv("WANDB_API_KEY")).getBytes(StandardCharsets.UTF_8));
        attributes = Map.of(
                "cookbook.language", "java",
                "cookbook.recipe", "12_run_evaluation",
                "cookbook.environment", getenv("COOKBOOK_ENVIRONMENT", "dev"));

        // 1) Look up the Evaluation Object + extract refs from its val.
        JsonNode evalObj = latestObject("recipe-11-eval-java");
        if (evalObj == null)
            fail("Evaluation Object `recipe-11-eval-java` not found. Run java/11_CreateEvaluation.java first.");
        String evalObjRef = "weave:///" + projectId + "/object/" + evalObj.get("object_id").asText() + ":" + evalObj.get("digest").asText();
        JsonNode ev = evalObj.get("val");
        String evaluateOpRef = ev.get("evaluate").asText();
        String predictAndScoreOpRef = ev.get("predict_and_score").asText();
        String summarizeOpRef = ev.get("summarize").asText();
        String scorerOpRef = ev.get("scorers").get(0).asText();
        String datasetRef = ev.get("dataset").asText();
        // The scorer Op's short_name (object_id) keys the per-row scores, the
        // wandb.runnable.* feedback_type, and the summarize/root output.
        String afterOp = scorerOpRef.substring(scorerOpRef.lastIndexOf("/op/") + "/op/".length());
        String scorerShortName = afterOp.split(":", 2)[0];
        System.out.println("Eval obj:  " + evalObj.get("object_id").asText() + " digest=" + evalObj.get("digest").asText().substring(0, 12) + "…");

        // 2) Look up the Model + its predict Op (recipe 08).
        JsonNode modelObj = latestObject("recipe-08-model-java");
        if (modelObj == null)
            fail("Model `recipe-08-model-java` not found. Run java/08_UseModel.java first.");
        String modelRef = "weave:///" + projectId + "/object/" + modelObj.get("object_id").asText() + ":" + modelObj.get("digest").asText();
        JsonNode modelPredictOp = latestObject("recipe-08-model-java.predict");
        if (modelPredictOp == null)
            fail("Model predict Op `recipe-08-model-java.predict` not found. Run java/08_UseModel.java first.");
        String modelPredictOpRef = "weave:///" + projectId + "/op/" + modelPredictOp.get("object_id").asText() + ":" + modelPredictOp.get("digest").asText();
        System.out.println("Model:     " + modelObj.get("object_id").asText() + " digest=" + modelObj.get("digest").asText().substring(0, 12) + "…");

        // 3) Walk the Dataset rows.
        Matcher dm = DATASET_REF.matcher(datasetRef);
        if (!dm.find()) fail("could not parse dataset_ref: " + datasetRef);
        String dsId = dm.group(1), dsDigest = dm.group(2);
        JsonNode dsMeta = get("/v2/" + entity + "/" + project + "/datasets/" + dsId + "/versions/" + dsDigest);
        String rowsRef = dsMeta.path("rows").asText();
        Matcher tm = TABLE_REF.matcher(rowsRef);
        String tableDigest = tm.find() ? tm.group(1) : rowsRef;
        JsonNode rowsRes = post("/table/query", Map.of("project_id", projectId, "digest", tableDigest));
        var rows = new ArrayList<JsonNode>();
        for (JsonNode rr : rowsRes.path("rows")) rows.add(rr.get("val"));
        System.out.println("Dataset:   " + dsId + " (" + rows.size() + " rows)");

        // 4) Build the 4-level Call trace.
        String displayName = "eval-java-" + Instant.now().getEpochSecond();
        String[] root = startCall(evaluateOpRef,
                Map.of("self", evalObjRef, "model", modelRef), null, null, displayName);
        String rootId = root[0], traceId = root[1];
        System.out.println("Root call: " + rootId + " (display_name=\"" + displayName + "\")");

        int nPass = 0, totalCalls = 1; // root
        for (JsonNode row : rows) {
            String[] ps = startCall(predictAndScoreOpRef,
                    Map.of("self", evalObjRef, "model", modelRef, "example", row), rootId, traceId, null);
            String psId = ps[0];

            // Predict child: invoke the (mocked) model — always returns expected.
            String[] pred = startCall(modelPredictOpRef,
                    Map.of("self", modelRef, "question", row.get("question").asText()), psId, traceId, null);
            String predId = pred[0];
            String prediction = row.get("answer").asText();
            endCall(predId, Map.of("answer", prediction));

            // Scorer child: compare prediction vs expected.
            String[] sc = startCall(scorerOpRef,
                    Map.of("output", prediction, "expected", row.get("answer").asText()), psId, traceId, null);
            String scId = sc[0];
            boolean score = prediction.equals(row.get("answer").asText());
            endCall(scId, score);

            // Link the score to the predict Call via a wandb.runnable.* Feedback
            // row (recipe 09's pattern) so the leaderboard attributes the scorer.
            post("/feedback/create", Map.of(
                    "project_id", projectId,
                    "weave_ref", "weave:///" + projectId + "/call/" + predId,
                    "feedback_type", "wandb.runnable." + scorerShortName,
                    "payload", Map.of("output", score),
                    "runnable_ref", scorerOpRef,
                    "call_ref", "weave:///" + projectId + "/call/" + scId));

            // End predict_and_score. The `scores` key MUST be the scorer Op's
            // short name so /eval_results/query links it to val.scorers.
            endCall(psId, Map.of(
                    "output", prediction,
                    "scores", Map.of(scorerShortName, score),
                    "model_latency", MODEL_LATENCY));

            if (score) nPass++;
            totalCalls += 3; // predict_and_score + predict + scorer
        }

        // Summarize: sibling of predict_and_score under the root.
        String[] sum = startCall(summarizeOpRef, Map.of("self", evalObjRef), rootId, traceId, null);
        double passRate = rows.isEmpty() ? 0.0 : (double) nPass / rows.size();
        // summarize.output AND root.output are keyed by the scorer short name
        // (matching val.scorers + per-row scorer_key) plus model_latency.mean.
        Map<String, Object> aggregatedOutput = Map.of(
                scorerShortName, Map.of("true_count", nPass, "true_fraction", passRate),
                "model_latency", Map.of("mean", MODEL_LATENCY));
        endCall(sum[0], aggregatedOutput);
        totalCalls++; // summarize

        // 5) End the root with the proper summary shape.
        post("/call/end", Map.of("end", Map.of(
                "project_id", projectId,
                "id", rootId,
                "ended_at", now(),
                "summary", Map.of(
                        "status_counts", Map.of("success", totalCalls, "error", 0),
                        "weave", Map.of("status", "success", "display_name", displayName)),
                "output", aggregatedOutput)));
        System.out.printf("Trace done: %d calls, pass_rate=%.2f%n", totalCalls, passRate);

        // --- verification ---
        // /eval_results/query with the root call_id aggregates per-row trial
        // data + scorer stats; evaluation_ref should match our Eval Object.
        Thread.sleep(2000);
        JsonNode results = null;
        for (int i = 0; i < 8; i++) {
            JsonNode r = post("/v2/" + entity + "/" + project + "/eval_results/query", Map.of(
                    "evaluation_call_ids", List.of(rootId),
                    "include_rows", true,
                    "include_summary", true));
            if (r.path("total_rows").asInt(-1) == rows.size()) {
                results = r;
                break;
            }
            Thread.sleep(1000);
        }
        if (results == null)
            fail("eval_results/query did not return " + rows.size() + " rows after 8 attempts");

        JsonNode evals = results.path("summary").path("evaluations");
        if (evals.size() != 1) fail("expected 1 evaluation in summary, got " + evals.size());
        JsonNode evSummary = evals.get(0);
        if (!evalObjRef.equals(evSummary.path("evaluation_ref").asText()))
            fail("evaluation_ref: " + evSummary.get("evaluation_ref"));
        var scorerKeys = new ArrayList<String>();
        for (JsonNode s : evSummary.path("scorer_stats")) scorerKeys.add(s.path("scorer_key").asText());
        if (!scorerKeys.contains(scorerShortName))
            fail("'" + scorerShortName + "' missing from scorer_stats: " + scorerKeys);
        System.out.println("Verified:  /eval_results/query returned " + results.path("total_rows").asInt()
                + " rows, evaluation_ref matches, scorer_stats=" + scorerKeys);
    }
}
